File: app/src/main/java/com/checkin/app/checkin/User/UserRepository.java
package com.checkin.app.checkin.User;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.checkin.app.checkin.Data.ApiClient;
import com.checkin.app.checkin.Data.ApiResponse;
import com.checkin.app.checkin.Data.BaseRepository;
import com.checkin.app.checkin.Data.NetworkBoundResource;
import com.checkin.app.checkin.Data.Resource;
import com.checkin.app.checkin.Data.RetrofitLiveData;
import com.checkin.app.checkin.Data.WebApiService;
import com.checkin.app.checkin.Misc.GenericDetailModel;
import com.checkin.app.checkin.Utility.ProgressRequestBody;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.File;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;

public class UserRepository extends BaseRepository {
    private static UserRepository INSTANCE = null;
    private final WebApiService mWebService;

    private UserRepository(Context context) {
        mWebService = ApiClient.Companion.getApiService(context);
    }

    public static UserRepository getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (UserRepository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new UserRepository(application.getApplicationContext());
                }
            }
        }
        return INSTANCE;
    }

    public LiveData<Resource<List<ShopCustomerModel>>> getUserRecentCheckins() {
        return new NetworkBoundResource<List<ShopCustomerModel>, List<ShopCustomerModel>>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<ShopCustomerModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.getUserRecentCheckins());
            }

            @Override
            protected void saveCallResult(List<ShopCustomerModel> data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<List<UserModel>>> getAllUsers() {
        return new NetworkBoundResource<List<UserModel>, List<UserModel>>() {

            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<UserModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.getUsers());
            }

            @Override
            protected void saveCallResult(List<UserModel> data) {
            }
        }.getAsLiveData();
    }

    /**
     * @param userPk - 0 for self, non-zero for others.
     * @return
     */
    public LiveData<Resource<UserModel>> getUser(long userPk) {
        return new NetworkBoundResource<UserModel, UserModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<UserModel>> createCall() {
                if (userPk == 0) {
                    return new RetrofitLiveData<>(mWebService.getPersonalUser());
                } else {
                    return new RetrofitLiveData<>(mWebService.getNonPersonalUser(userPk));
                }
            }

            @Override
            protected void saveCallResult(UserModel data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<UserModel>> postUserData(ObjectNode objectNode) {
        return new NetworkBoundResource<UserModel, UserModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<UserModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.postUserData(objectNode));
            }

            @Override
            protected void saveCallResult(UserModel data) {
            }
        }.getAsLiveData();
    }

    public Call<GenericDetailModel> postUserProfilePic(File pic, ProgressRequestBody.UploadCallbacks listener) {
        RequestBody requestFile = RequestBody.create(MediaType.parse("image/jpeg"), pic);
        ProgressRequestBody requestBody = new ProgressRequestBody(requestFile, listener);
        final MultipartBody.Part body = MultipartBody.Part.createFormData("profile_pic", "profile.jpg", requestBody);
        return mWebService.postUserProfilePic(body);
    }
}

File: app/src/main/java/com/checkin/app/checkin/user/UserRepository.kt
package com.checkin.app.checkin.user

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import com.checkin.app.checkin.data.BaseRepository
import com.checkin.app.checkin.data.network.ApiClient.Companion.getApiService
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.data.network.RetrofitLiveData
import com.checkin.app.checkin.data.network.WebApiService
import com.checkin.app.checkin.data.resource.NetworkBoundResource
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.misc.models.GenericDetailModel
import com.checkin.app.checkin.user.models.ShopCustomerModel
import com.checkin.app.checkin.user.models.UserLocationModel
import com.checkin.app.checkin.user.models.UserModel
import com.checkin.app.checkin.utility.ProgressRequestBody
import com.checkin.app.checkin.utility.ProgressRequestBody.UploadCallbacks
import com.checkin.app.checkin.utility.SingletonHolder
import com.fasterxml.jackson.databind.node.ObjectNode
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.Call
import java.io.File

class UserRepository private constructor(context: Context) : BaseRepository() {
    private val mWebService: WebApiService = getApiService(context)

    val userRecentCheckins: LiveData<Resource<List<ShopCustomerModel>>>
        get() = object : NetworkBoundResource<List<ShopCustomerModel>, List<ShopCustomerModel>>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<List<ShopCustomerModel>>> {
                return RetrofitLiveData(mWebService.userRecentCheckins)
            }
        }.asLiveData

    val allUsers: LiveData<Resource<List<UserModel>>>
        get() = object : NetworkBoundResource<List<UserModel>, List<UserModel>>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<List<UserModel>>> {
                return RetrofitLiveData(mWebService.users)
            }
        }.asLiveData

    /**
     * @param userPk - 0 for self, non-zero for others.
     * @return
     */
    fun getUser(userPk: Long): LiveData<Resource<UserModel>> {
        return object : NetworkBoundResource<UserModel, UserModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<UserModel>> {
                return if (userPk == 0L) {
                    RetrofitLiveData(mWebService.personalUser)
                } else {
                    RetrofitLiveData(mWebService.getNonPersonalUser(userPk))
                }
            }
        }.asLiveData
    }


    fun postUserData(objectNode: ObjectNode): LiveData<Resource<UserModel>> {
        return object : NetworkBoundResource<UserModel, UserModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<UserModel>> {
                return RetrofitLiveData(mWebService.postUserData(objectNode))
            }
        }.asLiveData
    }

    fun postUserProfilePic(pic: File, listener: UploadCallbacks<GenericDetailModel>): Call<GenericDetailModel> {
        val requestFile = RequestBody.create(MediaType.parse("image/jpeg"), pic)
        val requestBody = ProgressRequestBody(requestFile, listener)
        val body = MultipartBody.Part.createFormData("profile_pic", "profile.jpg", requestBody)
        return mWebService.postUserProfilePic(body)
    }

    fun postUserCurrentLocation(data: UserLocationModel): Call<UserLocationModel> {
        return mWebService.postUserLocation(data)
    }

    fun postUserReferralCode(data: ObjectNode): LiveData<Resource<UserModel>> {
        return object : NetworkBoundResource<UserModel, UserModel>() {
            override fun createCall(): LiveData<ApiResponse<UserModel>>? {
                return RetrofitLiveData(mWebService.postUserReferralCode(data))
            }
        }.asLiveData
    }

    companion object : SingletonHolder<UserRepository, Application>({ UserRepository(it.applicationContext) })
}